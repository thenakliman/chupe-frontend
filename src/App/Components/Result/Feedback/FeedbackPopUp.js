import React from 'react';
import propTypes from 'prop-types';

require('./FeedbackPopUp.css');

export class FeedbackPopUp extends React.Component {
  constructor(props) {
    super(props);
    this.state = {description: '', feedbackGivenToUser: ''};
    this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
    this.handleFeedbackTargetChange = this.handleFeedbackTargetChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  handleDescriptionChange(value) {
    this.setState({description: value});
  }

  handleFeedbackTargetChange(value) {
    this.setState({feedbackGivenToUser: value});
  }

  onSubmit() {
    this.props.onSubmit({
      description: this.state.description,
      givenTo: this.state.feedbackGivenToUser,
    });
  }

  render() {
    return (
        <div className={'feedback-pop-up-container'}>
          <div className={'feedback-pop-up-inner'}>
            <div>
              Feedback:
            </div>
            <textarea className={'feedback-description'}
                      value={this.state.feedback}
                      onChange={(event) => this.handleDescriptionChange(event.target.value)}
                      resize={'none'}/>

            Feedback To:
            <select value={this.state.feedbackGivenToUser}
                    className={'select-user-feedback'}
                    onChange={(event) => this.handleFeedbackTargetChange(event.target.value)}
            >
              <option key={''} value={''}> Select user</option>
              {
                this.props.users.map((user) =>
                    <option key={user.userName} value={user.userName}>
                      {user.userName}
                    </option>)
              }
            </select>
            <div className={'button-container'}>
              <button className={'feedback-buttons'}
                      id={'feedback-submit-button'}
                      onClick={this.onSubmit}>
                Submit
              </button>
              <button className={'feedback-buttons'}
                      id={'feedback-cancel-button'} onClick={this.props.onCancel}>
                Cancel
              </button>
            </div>
          </div>
        </div>
    );
  }
}

FeedbackPopUp.propTypes = {
  users: propTypes.array.isRequired,
  onSubmit: propTypes.func.isRequired,
  onCancel: propTypes.func.isRequired,
};
