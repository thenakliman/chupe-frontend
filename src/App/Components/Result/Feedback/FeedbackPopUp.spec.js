/* eslint-disable */
import React from 'react';
import {FeedbackPopUp} from './FeedbackPopUp';
/* eslint-enable */
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';


const users = [{
  userName: 'user - 1',
}, {
  userName: 'user - 2',
}];

describe('Show feedback pop up component', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(
      <FeedbackPopUp
        users={users}
        onSubmit={jest.fn()}
        onCancel={jest.fn()}
      />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have default description empty string', () => {
    const wrapper = shallow(
      <FeedbackPopUp
        users={users}
        onSubmit={jest.fn()}
        onCancel={jest.fn()}
      />);

    expect(wrapper.state().description).toEqual('');
  });

  it('should have default users empty string', () => {
    const wrapper = shallow(
      <FeedbackPopUp
        users={users}
        onSubmit={jest.fn()}
        onCancel={jest.fn()}
      />);

    expect(wrapper.state().feedbackGivenToUser).toEqual('');
  });

  it('should set state to newly changed description', () => {
    const wrapper = shallow(
      <FeedbackPopUp
        users={users}
        onSubmit={jest.fn()}
        onCancel={jest.fn()}
      />);

    wrapper.find('textarea').simulate('change',
        {target: {value: 'description-value'}});

    expect(wrapper.state().description).toEqual('description-value');
  });

  it('should have option for each user and default for select', () => {
    const wrapper = shallow(
      <FeedbackPopUp
        users={users}
        onSubmit={jest.fn()}
        onCancel={jest.fn()}
      />);

    expect(wrapper.find('option').length).toBe(3);
  });

  it('should set user on change of select', () => {
    const wrapper = shallow(
      <FeedbackPopUp
        users={users}
        onSubmit={jest.fn()}
        onCancel={jest.fn()}
      />);

     wrapper.find('select').simulate('change', {target: {value: 'user-1'}});

    expect(wrapper.state().feedbackGivenToUser) .toBe('user-1');
  });

  it('should call on submit button', () => {
    const onSubmit = jest.fn()
    const wrapper = shallow(
      <FeedbackPopUp
        users={users}
        onSubmit={onSubmit}
        onCancel={jest.fn()}
      />);

    wrapper.setState({description: 'fake', feedbackGivenToUser: 'user'});

    wrapper.find('#feedback-submit-button').simulate('click');
    expect(onSubmit).toHaveBeenCalledWith({
        description: 'fake', givenTo: 'user'});
  });

  it('should call on cancel button', () => {
    const onCancel = jest.fn()
    const wrapper = shallow(
      <FeedbackPopUp
        users={users}
        onSubmit={jest.fn()}
        onCancel={onCancel}
      />);

    const feedback = {
      description: 'fake',
      feedbackGivenToUser: 'user-1'
    };

    wrapper.setState(feedback);

    wrapper.find('#feedback-cancel-button').simulate('click');
    expect(onCancel).toHaveBeenCalledWith();
  });
});
