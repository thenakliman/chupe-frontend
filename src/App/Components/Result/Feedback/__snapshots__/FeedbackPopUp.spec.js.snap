// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show feedback pop up component should match snapshot 1`] = `
<div
  className="feedback-pop-up-container"
>
  <div
    className="feedback-pop-up-inner"
  >
    <div>
      Feedback:
    </div>
    <textarea
      className="feedback-description"
      onChange={[Function]}
      resize="none"
    />
    Feedback To:
    <select
      className="select-user-feedback"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
         Select user
      </option>
      <option
        key="user - 1"
        value="user - 1"
      >
        user - 1
      </option>
      <option
        key="user - 2"
        value="user - 2"
      >
        user - 2
      </option>
    </select>
    <div
      className="button-container"
    >
      <button
        className="feedback-buttons"
        id="feedback-submit-button"
        onClick={[Function]}
      >
        Submit
      </button>
      <button
        className="feedback-buttons"
        id="feedback-cancel-button"
        onClick={[MockFunction]}
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
